### DISCUSSION TITLE
Feature Suggestion

### DISCUSSION THREAD

**Discussion by ghost** [CombatExtended-Continued/CombatExtended#791] [General]
## Suggestions

Many things are suggested on our [discord](https://discord.gg/safyWuA) server. Every so often people ask if they ever get implemented or noticed by the development team and yes, they do but many times its counterpoints outweigh the benefits of said suggestion so they don't end up getting implemented and thrown into oblivion.

I'm opening this thread to house future suggestions and to clean up the issues tab.

**Comment by ghost** [comment]
# Gas Mask Toggle

With Combat Extended smoke mechanics, fighting--or even trying to escape from--even moderate sized fires in enclosed spaces can be quite dangerous in short order, as pawns are quickly overcome by smoke inhalation. As a result, it would be practical in many cases for a pawn to have a gas mask handy for emergencies--especially since pawns will currently wander into smoke filled areas. However, due to the stat impacts of wearing a gas mask, they're impractical to wear around. 

The solution, then, would be a UI gizmo only visible for pawns who have a gas mask. It would have three options:
- Disabled: Pawns will not wear a gas mask. If they're wearing one, they will store it in their inventory.
- Enabled: Pawns will don the gas mask in their inventory.
- Automatic: Pawns will automatically put on the gas mask in their inventory if they enter an area with hazardous levels of smoke. Once they're no longer in a hazardous area, they will remove the mask and put it back in their inventory (this should be a fairly low priority job, so we don't see colonists stopping in the middle of firefighting to take off their mask because the smoke on the tile they're standing on disappeared).

My thought would be that this could be controlled by an apparelTag, or something like that. It would also probably be worthwhile to have some functionality to ensure that pawns will choose the best gas mask in their inventory (e.g putting on a true Gas Mask rather than a Primitive Gas Mask), and to make sure they're not removing a piece of headgear that already completely protects them from smoke (like a power armor helmet) when in Automatic mode.

**Comment by ghost** [comment]
## Night Vision

The addition of a night vision comp with adjustable effectiveness values that could be added to weapons, apparel and implants which simply negates the aiming penalty CE introduces when pawns shoot at targets surrounded by darkness.

Example:

- No night vision comp: 100% penalty
- Crude photomultiplier tube: 50% penalty
- Military-grade night vision optics: 25% penalty
- [Full-color night vision optics](https://www.thedrive.com/the-war-zone/25803/this-is-what-color-night-vision-looks-like-and-it-stands-to-be-a-game-changer): 0% penalty, equivalent to daytime

XML Example:
```xml
<comps>
  <li Class="CombatExtended.CompProperties_NightVisionCE">
    <penaltyFactor>1</penaltyFactor>
    <penaltyFactor>0.50</penaltyFactor>
    <penaltyFactor>0.25</penaltyFactor>
    <penaltyFactor>0</penaltyFactor>
  </li>
</comps>
```

**Comment by ghost** [comment]
## Non-Colonist Loadout Framework

With the surge in player-controlled, non-colonists weapon users that can be equipped with ranged weapons, a common problem has arisen, these non-colonists usually don't have an inventory and currently there's no way to keep these pawns supplied with ammo for their weapons.

Most prominent mods in this area:

- Vanilla Factions Expanded - Mechanoid
- RimSec - Security
- What The Hack

Currently the most straightforward solution to this issue is to directly plug these pawns into the loadout system, allowing the player to manually control the ammo supply. At present, these pawns are generally not plugged into that system or to the related inventory jobs. Ideally, a new framework for this type of pawns could be developed with a simplified interface that can be applied to these pawns  to quickly make them compatible, rather than creating individual solutions for them.

**Comment by ghost** [comment]
## Improved Ammo System

- **Simplified Ammo:** Condense all pistol, rifle, and other calibers into simple generic ammo
  - Pistol, Shotgun, Rifle ammo
  - Light, Medium, Heavy ammo
  - Light Charge, Medium Charge, Heavy Charge ammo
- **No Ammo:** Allow players to swap between different ammo types FMJ / HP / AP / AP-I / AP-HE.

To be expanded (Feel free to edit).

**Reply by Alicecomma** [reply]
The best approach I figured is to set a 'SimplerAmmo' tag that is checked instead of the ammo itself in AmmoSetDef. Projectiles are unchanged, so the same ammo object can be used for multiple ammo types. Originally I made the SimplerAmmo refer to other existing ammo instead of a generic ammo object, making the ammo system _simpler_ rather than completely without character.

**Comment by Alicecomma** [comment]
# Weapon parts

# MOVED TO #948

**Reply by Alicecomma** [reply]
## Modifying UI
### Current state
<details>
<summary>
OPEN
</summary>

![image](https://user-images.githubusercontent.com/4662406/128790541-7c19f0fb-b3c6-4220-918f-b0e12a075a03.png)

- Ground item not updated

![image](https://user-images.githubusercontent.com/4662406/128788906-c9a41ca0-e871-4884-ac86-8f56477ddbe9.png)

- Inventory icon not updated

<img src="https://user-images.githubusercontent.com/4662406/128789068-0168d817-9092-40c7-b707-a12137f0a33e.png" height="400" />
</details>

### What is an attachment item?

Multiple options:
1. Completely individual items with/without quality (weapon-like), currently also unstackable
2. Generic attachments (scope tier 1, muzzle tier 3) -- Makes 'mixing' attachments hard
3. Generic items (components, wood log) -- Makes 'mixing' attachments hard

What is the goal?
1. Make attachments the new 'ammo system' level of micromanagement, goal to try and allow mixing between guns
2. Make attachments more similar to components, allow with less micro
3. Attachments are common and easy to change, allow minimal micro (also minimal character?)

### What should the UI look like?
<details>
<summary>
OPEN
</summary>

A: Make sure it looks similar to RimWorld. Existing menus are those for choosing hairstyles - ideology styling tab. Showing individual gun parts perhaps like body parts.
<img src="https://user-images.githubusercontent.com/4662406/127488494-05b60b2b-7048-42f0-9737-c3a80b387a2a.png" height="200" />
- [1:58 AM] Andross: Fucking STALKER be like
<img src="https://user-images.githubusercontent.com/4662406/127489274-f281dad5-a6a8-4937-9976-19731888c5a6.png" height="200" />
Latest:
<img src="https://cdn.discordapp.com/attachments/866794648102371339/872295927577182248/unknown.png" height="400" />
</details>

### What are UI items?
<details>
<summary>
OPEN
</summary>

1.: The ThingDef of the attachment.
- FOR INVISIBLE ATTACHMENTS (barrel)
- Problem: only few exist - gets boring, not descriptive especially if all scopes fit on all weapons at some point.

2.: The attachment Graphic alone.
- Problem: no idea where the graphic should be.
- Solution: could use blit cropping.

3.: The current gun with the attachment in correct position.
Current:
<img src="https://user-images.githubusercontent.com/4662406/128103098-1e0609e9-9a13-4dc0-99c6-e64c58e404de.png" height="200" />

4.: MenuIcons, abstract representation of the attachment.
- Problem: more artist work. Not extremely consistent with RW UI style.

5.: Texts with the ThingDef of the attachment.
- Works for essentially all attachments
Current:
<img src="https://user-images.githubusercontent.com/4662406/128106216-633bdf4a-9fcf-4742-91f7-425e810fd232.png" height="200" />
</details>

**Reply by Alicecomma** [reply]
## Rendering
### How to render attachments over existing weapons?
A: Paste black outlines behind, colour in front to add to existing weapons
<img src="https://user-images.githubusercontent.com/4662406/127488956-6dc01ad3-19bc-4b0b-b499-60f447624a12.png" height="300" />
Current iteration by Karim:
<img src="https://user-images.githubusercontent.com/4662406/127489350-a6a8d5b3-5183-42df-9aee-075eb9d7b0c2.png" height="100" />

What graphics need be supplied? In current iteration, these:
<img src="https://user-images.githubusercontent.com/4662406/127489406-20948cfa-eef3-4c62-8d99-8de491364bfe.png" height="200" />

**Reply by Alicecomma** [reply]
## Visual attachment points
### Where do the attachments visually attach?
**Answer 1 (Alistaire)**: On attachments points Vector2 assigned in def (similar to Mrofa hands). Specifically use of 'attachment point' that is on the gun!
- [2:17 PM] NumberOfImposterAttempts: XML supplies pixel to be matched by attachmentTag, C# moves parts to match pixels
- [2:20 PM] /S/afairette: different attachments would draw better in front or behind the main body

<img src="https://user-images.githubusercontent.com/4662406/127491141-f1c9e82e-4c8a-4985-a1c3-016f0450ed87.png" width="300" />

**BENEFITS**: anything could attach anywhere; large amount of content added with only a single attachment XML; no need for aligning sprites manually.
**DOWNSIDES**:
- [1:35 AM] Karim: do you want to do every offset [red.: XML] in the world?

<hr>

**Answer 2 (Karim)**:
- [1:34 AM] Karim: the are pre aligned, as in the attachment is made for the weapon. you just export it.

**BENEFITS**: less work needed for XML team per image
**DOWNSIDES**: more work needed for graphics team; likely impossible to mix-and-match parts; Fallout 4-like approach.

<hr>

### What about guns that do not have the visuals ready for the part system?
**A**: Assume the whole gun is a 'part' using a Comp that has simpler settings (attachment points etc.) so a mod author could also implement the system without needing to redo graphics.

**Reply by Alicecomma** [reply]
## Map items
### What would the gun part items look like when they're not on a gun?
A: Boxes similar to bodyparts, possibly with the right type of gun part icon on it

### What truly is a scuffed weapon?
A: Make it like the ThingDef for crafted items in progress.

**Reply by Alicecomma** [reply]
## Codebase clashes
- [2:19 PM] NumberOfImposterAttempts: Karim suggested CompAmmoUser might not be necessary anymore with the system, but lots of the codebase refers to that comp

**Reply by Alicecomma** [reply]
## XML
### What is the XML structure?
- [8:45 AM] Andross: Then maybe the gun could have something like
```
<attachmentPoints>
  <li>Stock<li>
  <li>Optic</l>
  <li>Underbarrel</li>
  <li>Magazine</li>
</attachmentPoints>
```

- [8:39 AM] Andross: XML concept
![image](https://user-images.githubusercontent.com/4662406/127494031-3233684b-e4ca-4f53-89cc-3cc5a5313853.png)

- [7:53 AM] Andross: How do i define the offsets for that gun
![image](https://user-images.githubusercontent.com/4662406/127494151-df9c3603-ce87-413f-9f65-4d18a62e7596.png)

![image](https://user-images.githubusercontent.com/4662406/127494278-dae17038-65ec-43ce-a238-19dc9ae10c4a.png)

**Reply by ghost** [reply]
## Escapen't From Tarkov

### How To Acquire

#### TRADERS
**Combat Suppliers** - Carry a small assortment of attachments (Low - Normal quality [Maybe very rare high quality]).

**Orbital Combat Suppliers** - Carrying large quantities of attachments (Normal+ quality)

#### LOOTING
Guns taken from raiders can be looted for attachments, these usually have taken damage during the fight reducing their bonuses.

#### RESEARCH
Attachments must be unlocked through research, machining being a prerequisite.

**Model A -** Separate attachments into three tiers:
 - **Tier 1 - Basic Attachments** 
   Simple foregrips, shoddy scopes, basic shotgun / grenade launcer, magazines and some muzzle stuff.
 - **Tier 2  - Advanced Attachments**
   Red Dots / Holographic sights, improved muzzle stuff, shotgun / grenade launcher, better scopes, flashlights and lasers.
 - **Tier 3 - Spacertech Attachments**
   Night vision scopes, plasma(?) launchers, etc.

**Model B -** Separate each attachment type into it's own tier / tiers:
- **Tier 1 Optics:**
   Unlocks shoddy scopes and foregrips.
  ***Prerequisites:*** *Machining / Gunsmithing.*
- **Tier 2 Optics:**
  Unlocks x2/3 augment scopes and red dots.
  ***Prerequisites:*** *Machining / Gunsmithing / T1 Optics.*
- **Tier 3 Optics:**
  Unlocks x4+ augment scopes and holos.
 ***Prerequisites:*** *Machining / Gunsmithing / T2 Optics / Microelectronics.*
- **Tier 4 Optics:**
  Unlocks night vision scopes and spacer optics.
   ***Prerequisites:*** *Machining / Gunsmithing / T3 Optics / Microelectronics / Multi-analyzer.*

Then after finishing the desired research, these can be crafted in the **Gunsmithing Bench** unlocked after researching Machining.

**Recommendation:** Add a new research tab explicitly for attachments.

### Attachment Types

**OPTICS**
- Red Dot
- Holographic
- Scopes

#### STOCKS
Stocks reduce recoil in general, can also increase aim time depending on the recoil reduction.

#### MAGAZINES
- Extended Magazines
- Drum Magazines
   Have the advantage of holding large amounts of ammo (50, 75, 100 etc) while greatly increasing the bulk and weight of a gun
- Stackable Magazines [Faster reload]

#### UNDERBARREL
- **Foregrips**
  - Reduces recoil
  - Reduces sway
- **Bipods**
  - Reduces recoil
  - Reduces sway
  - Increases weapon bulk

#### SIDERAIL
- **Flashlight**
  - Night vision efficiency bonus
- **Laser Sight**
   - Reduces sway

#### MUZZLE
- **Compensators**
   - Reduces recoil
- **Flash Hiders**
   - Reduces muzzle flash
   - Reduces recoil
- **Supressors**
   - Reduces muzzle flash
   - Increases weapon bulk
   - Hunting stealth bonus
- **Chokes**
   - Reduces spread

**Reply by Alicecomma** [reply]
## Crafting/Adding Attachments
### How is the UI accessed?

Option 1: Dedicated crafting table.
- Open UI much like ammo crafting table
- Problem: How would you select each weapon? Show all on the map? Show all pawns? Only allow opening UI with pawn there - like hairdresser?

Option 2: Can add anywhere from inventory.
- Open UI right-click 'Change attachments' or so.
- Starting job, takes X seconds exactly like reloading does.

**Comment by Alicecomma** [comment]
# Smoke/Soot (#75)

Soot
- Start to dissipate into filth once the source of the flames has been removed.
- Make all smoke dissipate into filth, rather than only very low densities.
- Smoke particles tend to be sticky due to the tannins and tar in them.  Once the colloid has settled, it generally becoems a hard-to-remove sticky mess.
-  There's also the ashy stuff that's super flimsy and jumps up.

A/C
- Separate air cleaner with a filter for dealing with heavy smoke
- Don't know if it's worth making the AC break down.  Bit of a hard sell that cleaning smoke should consume components.
- Filter insert
- Could add a filter option to ACs that would either:
1.- Increase their breakdown chance as it clears smoke
2.- A "filter" that needs to be cleared
